In [3]:
import numpy as np
import pandas as pd


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [66]:

#TASK 1

df = pd.read_csv('/content/drive/MyDrive/adult.csv')

print ("HEAD OF DATA\n")
print(df.head(5))

print ("\nINFORMATION OF DATA\n")
print(df.info())

print("\nCHECKING NULL\n")
print (df.isnull().sum())

print("\nAfter replacing\n")
df.replace("?", np.nan, inplace=True)

print("\nRechecking\n")
print(df.isnull().sum())

print(df.shape)

HEAD OF DATA

   age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18 

In [48]:
#TASK 2

print("Count of missing values per column\n")
print(df.isnull().sum().sort_values(ascending=False))

print("\nPercentage of missing values per column\n")
print((df.isnull().mean()*100).sort_values(ascending=False))

Count of missing values per column

occupation        1843
workclass         1836
native.country     583
fnlwgt               0
education            0
education.num        0
age                  0
marital.status       0
relationship         0
sex                  0
race                 0
capital.gain         0
capital.loss         0
hours.per.week       0
income               0
dtype: int64

Percentage of missing values per column

occupation        5.660146
workclass         5.638647
native.country    1.790486
fnlwgt            0.000000
education         0.000000
education.num     0.000000
age               0.000000
marital.status    0.000000
relationship      0.000000
sex               0.000000
race              0.000000
capital.gain      0.000000
capital.loss      0.000000
hours.per.week    0.000000
income            0.000000
dtype: float64


In [49]:
#TASK 3

df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')


age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64
age                  0
workclass            0
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64
   age workclass  fnlwgt     education  education.num marital.status  \
0   90   Private   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66   Private  186061  Some-college             10        Widowed

/tmp/ipykernel_685/1962153340.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['workclass'].fillna(df['workclass'].mode()[0], inplace=True)


**Justification**:



I replaced the missing values in occupation with Unknown because we cannot determine the actual occupation, and using the mode could assign an incorrect occupation.

I also replaced the missing values in workclass with Unknown for the same reason. This keeps all the records in the dataset without making assumptions about the missing values.

There were no missing values in the numerical columns, so no numerical imputation was required.


In [70]:
#TASK 4

print("Exact duplicates\n")
print(df.duplicated().sum())

print("Duplicates ignoring income:", df.drop(columns=['income']).duplicated().sum())



Exact duplicates

24
Duplicates ignoring income: 25


In [51]:
#TASK 5

print("Unique in education:", df['education'].unique())
print("Unique in marital-status:", df['marital.status'].unique())
print("Unique in natove-country:", df['native.country'].unique())

df['education'].str.startswith(' ')
df['education'].str.endswith(' ')


df['marital.status'].str.startswith(' ')
df['marital.status'].str.endswith(' ')


df['native.country'].str.startswith(' ')
df['native.country'].str.endswith(' ')

#no string found with trailing white spaces




Unique in education: ['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']
Unique in marital-status: ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']
Unique in natove-country: ['United-States' nan 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


,native.country
0,False
1,False
2,False
3,False
4,False
...,...
32556,False
32557,False
32558,False
32559,False


In [52]:
#TASK 6

print(df.describe())
print(df.describe(include="object"))


                age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  
       workclass education      marital.status      occupation 

the mean age is 38.5 which tells us the average age of the adults in this dataset. it shows us what is usually the age of people for who we have collected the data.

the media age is 37 which tells us that the half of the people's age lies above that and for other half it lies below that mark.


the mean hours-per-week is 40.43 which shows on average how much the people in the dataset work.

the median hours-per-week is 40 which is approximately the same as the mean, showing that it is normal distribution.

In [58]:
#TASK 7
print("Most common Occupation", df['occupation'].value_counts())
print("Percentage dist of sex",df['sex'].value_counts(normalize=True)*100)
print(df['income'].value_counts(normalize=True) * 100)


Most common Occupation occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: count, dtype: int64
Percentage dist of sex sex
Male      66.920549
Female    33.079451
Name: proportion, dtype: float64
income
<=50K    75.919044
>50K     24.080956
Name: proportion, dtype: float64


the income classes are not balanced. the percentage distributions shows that the class <=50k is heavier and the class >50k is lighter.

In [71]:
#Task 8

df.groupby('education')['education.num'].unique()

,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]


In [72]:
#TASK 9

import pandas as pd
import sqlite3


df_original = pd.read_csv('/content/drive/MyDrive/adult.csv')

print("Original CSV shape:", df_original.shape)


conn = sqlite3.connect("adult_income.db")


df_original.to_sql(
    "adult_income",
    conn,
    if_exists="replace",
    index=False
)


query = """
SELECT *
FROM adult_income
WHERE age > 30
"""

df_sql = pd.read_sql_query(query, conn)


df_pandas = df_original[df_original["age"] > 30]


print("\n--- Shape Verification ---")
print("Original CSV shape:", df_original.shape)
print("SQL result shape:", df_sql.shape)
print("Pandas filtered shape:", df_pandas.shape)

print("\n--- Condition Verification ---")
print("All extracted ages > 30:", (df_sql["age"] > 30).all())
print("Minimum extracted age:", df_sql["age"].min())


df_sql_sorted = df_sql.sort_values(
    by=list(df_sql.columns)
).reset_index(drop=True)

df_pandas_sorted = df_pandas.sort_values(
    by=list(df_pandas.columns)
).reset_index(drop=True)

print("\n--- Data Verification ---")
print("SQL and pandas results are identical:",
      df_sql_sorted.equals(df_pandas_sorted))


print("\n--- Extracted Data ---")
print(df_sql.head())


conn.close()

Original CSV shape: (32561, 15)

--- Shape Verification ---
Original CSV shape: (32561, 15)
SQL result shape: (21989, 15)
Pandas filtered shape: (21989, 15)

--- Condition Verification ---
All extracted ages > 30: True
Minimum extracted age: 31

--- Data Verification ---
SQL and pandas results are identical: True

--- Extracted Data ---
   age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?    

**task 10**

dataset overview: rows:32561 cols:15 represents the data of adults and their income and other factors which play a role in understanding this data such as occupation, age, work class, etc.
Data quality issues found: there were missing values in occupation, workclass, and native country. there were 24 duplicated with involving the target class and 25 without the income class because even if other attributes ares same the incme mayor may not be same. there were no trailign white spaces in the data. also no inconsistent mappings were found in the education and education-num.
Key observations: the income classes are imbalanced with more in <=50k. and the most common occupation is prof-speciality.
Cleaning decisions made :i replaced with unknown catgeory in both occupation and workclass because it makes sure no invalid data is added.